### PR TITLE
feat(query): Label Replace function support

### PR DESCRIPTION
**Pull request by TanviBhavsar**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

Support for prometheus label_replace function

### CODE REVIEW

**Comment by vishramachandran** [comment]
@TanviBhavsar Label Replace is not an instant function - it does not deal with instant data at all. So we need not mix it up with Instant Functions.

We just need a separate RangeVectorTransformer. Let us discuss in person and clarify our understanding.

File: core/src/main/scala/filodb.core/query/RangeVector.scala [hunk 1]
```diff
@@ -61,6 +61,11 @@ final case class CustomRangeVectorKey(labelValues: Map[UTF8Str, UTF8Str]) extend
   val sourceShards: Seq[Int] = Nil
 }
 
+final case class CustomRangeVectorKeyWithShards(labelValues: Map[UTF8Str, UTF8Str], sourceShards: Seq[Int])
```
Comment (vishramachandran) [inline core/src/main/scala/filodb.core/query/RangeVector.scala]: Why was this needed? I think you can simply use CustomRangeVectorKey

File: query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  override def apply(rangeVectorKey: RangeVectorKey): RangeVectorKey = {
+    val dstLabel = funcParams(0).asInstanceOf[String]
+    val replacementString = funcParams(1).asInstanceOf[String]
+    val srcLabel = funcParams(2).asInstanceOf[String]
+    try {
+      val regex = funcParams(3).asInstanceOf[String].r
+      val value = rangeVectorKey.labelValues.get(ZeroCopyUTF8String(srcLabel))
+
+      if (value.isDefined) {
+        val labelReplaceValue = regex.replaceAllIn(value.get.toString, replacementString)
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala]: As discussed in person, this api doest do what we want to do here. Need to find the right scala api, or use the java api.

File: query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  def apply(function: MiscellaneousFunctionId, funcParams: Seq[Any]): LabelTypeInstantFunction =
+  {
+    function match {
+      case LabelReplace       => LabelReplaceImpl(funcParams)
+      case _                  => throw new UnsupportedOperationException(s"$function not supported.")
+    }
+  }
+}
+
+case class LabelReplaceImpl(funcParams: Seq[Any]) extends LabelTypeInstantFunction {
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala]: you could simply have this as a function within LabelFunctions object. Any specific need for separate class?

File: query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  /**
+    * Validate the function before invoking the function.
+    */
+  require(funcParams.size == 4,
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala]: These validations should be within the RangeVectorTransformer which is instantiated at query planning time. Errors should ideally be reported much earlier before query processing starts.

File: coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala [hunk 5]
```diff
@@ -267,6 +269,19 @@ class QueryEngine(dataset: Dataset,
...
   }
 
+  private def materializeApplyMiscellaneousFunction(queryId: String,
+                                              submitTime: Long,
+                                              options: QueryOptions,
+                                              lp: ApplyMiscellaneousFunction): Seq[ExecPlan] = {
+    val vectors = walkLogicalPlanTree(lp.vectors, queryId, submitTime, options)
+    lp.function match {
+      case LabelReplace =>
+        vectors.foreach(_.addRangeVectorTransformer(exec.LabelFunctionMapper(lp.function,lp.functionArgs)))
```
Comment (vishramachandran) [inline coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala]: Consider having one RangeVectorTransformer with all the misc functions. The transformer (aka mapper) could perform the validation and apply of all functions.

File: query/src/test/scala/filodb/query/exec/rangefn/LabelFunctionSpec.scala [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        override def rows: Iterator[RowReader] = Seq(
+          new TransientRow(3L, 100d),
+          new TransientRow(4L, 200d)).iterator
+      })
+
+    val expectedLabels = List(Map(ZeroCopyUTF8String("instance") -> ZeroCopyUTF8String("demo.io new Label Value 90"),
+      ZeroCopyUTF8String("job") -> ZeroCopyUTF8String("test")),
+      Map(ZeroCopyUTF8String("instance") -> ZeroCopyUTF8String("demo.io:8080"))) // will not be replaced
+
+    val funcParams = Seq("instance", "$1 new Label Value $2", "instance", "(.*):90(.*)")
```
Comment (vishramachandran) [inline query/src/test/scala/filodb/query/exec/rangefn/LabelFunctionSpec.scala]: There is one more test case... What if there are just two groups and I use $3 ?

File: query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    "Cannot use LabelReplace without function parameters: " +
+      "instant-vector, dst_label string, replacement string, src_label string,regex string")
+
+  override def apply(rangeVectorKey: RangeVectorKey): RangeVectorKey = {
+    val dstLabel = funcParams(0).asInstanceOf[String]
+    val replacementString = funcParams(1).asInstanceOf[String]
+    val srcLabel = funcParams(2).asInstanceOf[String]
+    try {
+      val regex = funcParams(3).asInstanceOf[String].r
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala]: Move this out as a class member and ensure it runs during query planning. Regex validations will be caught there.

**Comment by TanviBhavsar** [comment]
> Excellent progress in understanding the query processing. I am very glad you are becoming more and more equipped to address all the query features.

Thanks Vish :)

File: query/src/main/scala/filodb/query/PlanEnums.scala [hunk 8]
```diff
@@ -204,6 +180,37 @@ object Cardinality extends Enum[Cardinality] {
...
+sealed abstract class MiscellaneousFunctionId(override val entryName: String) extends EnumEntry
+
+object MiscellaneousFunctionId extends Enum[MiscellaneousFunctionId] {
+  val values = findValues
+
+  case object DaysInMonth extends MiscellaneousFunctionId("days_in_month")
+
+  case object DaysOfMonth extends MiscellaneousFunctionId("day_of_month")
+
+  case object DayOfWeek extends MiscellaneousFunctionId("day_of_Week")
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/PlanEnums.scala]: Done

File: query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala [hunk 9]
```diff
@@ -123,3 +124,21 @@ final case class ScalarOperationMapper(operator: BinaryOperator,
 
   // TODO all operation defs go here and get invoked from mapRangeVector
 }
+
+final case class LabelFunctionMapper(function: MiscellaneousFunctionId,
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala]: Added MiscellaneousFunctionMapper

File: core/src/main/scala/filodb.core/query/RangeVector.scala (hunk 1 above)
Comment (TanviBhavsar) [reply core/src/main/scala/filodb.core/query/RangeVector.scala]: Removed CustomRangeVectorKeyWithShards and modified CustomRangeVectorKey to have shards as input

File: query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala (hunk 2 above)
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala]: Used Java API

File: query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala (hunk 3 above)
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala]: Changed the design

File: query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala (hunk 4 above)
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala]: Added validator to do all validations

File: coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala (hunk 5 above)
Comment (TanviBhavsar) [reply coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala]: Done

File: query/src/test/scala/filodb/query/exec/rangefn/LabelFunctionSpec.scala (hunk 6 above)
Comment (TanviBhavsar) [reply query/src/test/scala/filodb/query/exec/rangefn/LabelFunctionSpec.scala]: Added it in test case "should works with multiple groups and remove groups which do exist from the replacement string"

File: core/src/main/scala/filodb.core/query/RangeVector.scala (hunk 1 above)
Comment (vishramachandran) [reply core/src/main/scala/filodb.core/query/RangeVector.scala]: Ah - what I meant is that shards should not be part of the key. Why is this needed in the first place ?

File: core/src/main/scala/filodb.core/query/RangeVector.scala [hunk 10]
```diff
@@ -57,8 +57,8 @@ final case class PartitionRangeVectorKey(partBase: Array[Byte],
   override def toString: String = s"/shard:$sourceShard/${partSchema.stringify(partBase, partOffset)} [grp$groupNum]"
 }
 
-final case class CustomRangeVectorKey(labelValues: Map[UTF8Str, UTF8Str]) extends RangeVectorKey {
-  val sourceShards: Seq[Int] = Nil
+final case class CustomRangeVectorKey(labelValues: Map[UTF8Str, UTF8Str], sourceShards: Seq[Int] = Nil)
```
Comment (vishramachandran) [inline core/src/main/scala/filodb.core/query/RangeVector.scala]: Should not be added shards to the key. Lets discuss more in person.

File: query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+  def execute(): Observable[RangeVector]
+}
+
+case class LabelReplaceFunction(source: Observable[RangeVector], funcParams: Seq[Any])
+  extends MiscellaneousFunction {
+
+  val labelIdentifier: String= "[a-zA-Z_][a-zA-Z0-9_:\\-\\.]*"
+
+  private def pattern(): Option[Pattern] = {
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: I would make this a val instead of def, have it return an PAttern instead of an Option.
Throw any PatternSyntaxException out so it becomes a validation exception at construction time.
Add any other "require" inline outside of any method.
This way, you can remove the validator method from trait too.

File: query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    val regex = funcParams(3).asInstanceOf[String]
+    try {
+      val pattern: Pattern = Pattern.compile(regex)
+      Some(pattern)
+    }
+    catch {
+      case ex: PatternSyntaxException => {
+        require(false, "Invalid Regular Expression")
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: If PatternSyntaxException has more details of the regex error, it may be useful for user to know. You could simply throw that exception.

File: core/src/main/scala/filodb.core/query/RangeVector.scala (hunk 1 above)
Comment (TanviBhavsar) [reply core/src/main/scala/filodb.core/query/RangeVector.scala]: I could not just update labelValues of RangeVectorKey since it is a val. So I created new RangeVectorKey with the updated labels. If I use the original CustomRangeVectorKey, sourceShards will be nil.
So in labelReplaceImpl I create updated RangeVectorKey using: CustomRangeVectorKey(rangeVectorKey.labelValues.
         updated(ZeroCopyUTF8String(dstLabel), ZeroCopyUTF8String(labelReplaceValue)), rangeVectorKey.sourceShards)

File: query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala [hunk 13]
```diff
@@ -168,3 +174,22 @@ final case class ScalarOperationMapper(operator: BinaryOperator,
 
   // TODO all operation defs go here and get invoked from mapRangeVector
 }
+
+final case class MiscellaneousFunctionMapper(function: MiscellaneousFunctionId,
```
[the remark this answers is not in the document]
Comment (vishramachandran) [reply query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala]: @velvia Your suggestion makes sense, but every misc function does not follow the mapRV pattern. Among the misc functions only one or two do.

@TanviBhavsar has already done a grouping of the "misc" promQL functions that don't fall into sub-groups with  vastly reusable patterns. 

See `MiscellaneousFunctionId` for the full list.

File: query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala (hunk 11 above)
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: Made the change

File: query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala (hunk 12 above)
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: throwing PatternSyntaxException

File: coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala [hunk 14]
```diff
@@ -261,6 +262,15 @@ class QueryEngine(dataset: Dataset,
     }
   }
 
+  private def materializeApplyMiscellaneousFunction(queryId: String,
+                                              submitTime: Long,
+                                              options: QueryOptions,
+                                              lp: ApplyMiscellaneousFunction): Seq[ExecPlan] = {
+    val vectors = walkLogicalPlanTree(lp.vectors, queryId, submitTime, options)
+    vectors.foreach(_.addRangeVectorTransformer(exec.MiscellaneousFunctionMapper(lp.function,lp.functionArgs)))
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala]: Added space

File: query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala [hunk 15]
```diff
@@ -6,10 +6,16 @@ import filodb.core.metadata.Column.ColumnType
...
 import filodb.core.query._
 import filodb.memory.format.RowReader
-import filodb.query.{BinaryOperator, InstantFunctionId, QueryConfig}
-import filodb.query.exec.binaryOp.BinaryOperatorFunction
-import filodb.query.exec.rangefn.{DoubleInstantFunction, HistToDoubleIFunction, InstantFunction}
+import filodb.query.{BinaryOperator, InstantFunctionId, MiscellaneousFunctionId, QueryConfig}
 import filodb.query.InstantFunctionId.HistogramQuantile
+import filodb.query.MiscellaneousFunctionId.LabelReplace
+import filodb.query.exec.binaryOp.BinaryOperatorFunction
+import filodb.query.exec.rangefn.{
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala]: Used wildcard

File: query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala [hunk 16]
```diff
@@ -168,3 +174,22 @@ final case class ScalarOperationMapper(operator: BinaryOperator,
...
+  protected[exec] def args: String =
+    s"function=$function, funcParams=$funcParams"
+
+  val miscFunction: MiscellaneousFunction = {
+    function match {
+      case LabelReplace => LabelReplaceFunction(funcParams)
+      case _ => throw new UnsupportedOperationException(s"$function not supported.")
+    }
+  }
+  def apply(source: Observable[RangeVector],
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala]: Added line

File: query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+import filodb.memory.format.ZeroCopyUTF8String
+
+trait MiscellaneousFunction {
+  def execute(source: Observable[RangeVector]): Observable[RangeVector]
+}
+
+case class LabelReplaceFunction(funcParams: Seq[Any])
+  extends MiscellaneousFunction {
+
+  val labelIdentifier: String = "[a-zA-Z_][a-zA-Z0-9_:\\-\\.]*"
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: To keep label name consistent, I used the label regex from filodb/prometheus/parse/Parser.scala

File: query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  def labelReplaceImpl(rangeVectorKey: RangeVectorKey, funcParams: Seq[Any]): RangeVectorKey = {
+
+    val value: ZeroCopyUTF8String = if (rangeVectorKey.labelValues.contains(ZeroCopyUTF8String(srcLabel))) {
+      rangeVectorKey.labelValues.get(ZeroCopyUTF8String(srcLabel)).get
+    }
+    else {
+      // Assign dummy value as label_replace should overwrite destination label if the source label is empty but matched
+      ZeroCopyUTF8String("")
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: Good suggestion! Changed it to ZeroCopyUTF8String.empty.

File: query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+
+  require(dstLabel.matches(labelIdentifier), "Invalid destination label name")
+
+  try {
+    Pattern.compile(regexString)
+  }
+  catch {
+    case ex: PatternSyntaxException => {
+      throw new PatternSyntaxException("Invalid Regular Expression for label_replace\n" +
+        ex.getDescription, regexString, ex.getIndex)
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: Dont lose the original exception. It may be simplest to do
```
throw new IllegalArgumentException("Invalid Regular Expression for label_replace", ex)
```

Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: Done!
